### DISCUSSION TITLE
Mixing up titles within an album

### DISCUSSION THREAD

**Discussion by guerda** [beetbox/beets#5296] [General]
If the track number is not filled, beets (2.0 but also the previous version) confuses some titles:
Here's an example: Both files have the track number in their file name, but not in the track # tag.
beets now wants to (incorrectly) rename/tag one file to another one which exists and then name the correct one wrong, too.

     ≠ (#0) deadmau5 - where’s the drop? - superbia (ov) (2:22) -> (#5) avaritia (ov) (2:58)
     ≠ (#0) 5. avaritia (ov) (1:20) -> (#14) caritas (ov) (1:38)

Looks like this:
![image](https://github.com/beetbox/beets/assets/230782/c91810b8-ca0f-451b-bcd9-d5e12da1152e)

In sum, most of the files would be correctly tagged, but this would change the titles for worse.
Is there any idea how to solve this? I have the `fromfilename` plugin active.

**Comment by guerda** [comment]
It's even worse with some other example: There even the track number is correct and maintained, but still beets wants to change it to a different track.
Any idea how to change this? 

![image](https://github.com/beetbox/beets/assets/230782/480e302d-0d06-491b-896c-ed2cff47cb48)

**Comment by DjSlash** [comment]
Looks like you have files with bad/non-existing metadata. Beets is powerful, but this makes it hard to match correctly.

So, your options are:
- pre-condition the files, it can help to give them proper filenames and/or clean up tags, maybe (re)move extraneous files out of the way
- use the [chroma plugin](https://beets.readthedocs.io/en/stable/plugins/chroma.html); it fingerprints files and gives beets an extra identification power
- use the [edit plugin](https://beets.readthedocs.io/en/stable/plugins/edit.html); during an import session it gives you options to edit the metadata that's used for matching.

**Comment by guerda** [comment]
While I agree, this is not the case in the second case. Both ID3 tags and file names match and are correct. The track # is also correct, but still beets wants to edit it (for worse).
I am using the chroma plugin already for the really hard cases. Edit plugin might be interesting for me though.

**Comment by DjSlash** [comment]
Well, I'm not sure what's going on in your second case, but all in all it seems _something_ is off. Because it also shows 11 unmatched tracks. I wonder why.

**Comment by snejus** [comment]
@guerda could you install `mutagen` (Python) and show the output of `mid3v2 <your-file>`?

I have had a look at your **Deadmau5** import, see a couple of notes below:


#### Issues
Notice that `beets` correctly matched 11 of your tracks, and failed to match 4.

1. Files are missing for 3 titles:

- **ira (ov)**
- **fn pig (ov)**
- **caritas (ov)**

2. Each **monophobia (ov)** and **superbia (ov)** are duplicated across two files:

- `6. monophobia (ov)` and `deadmau5 - where's the drop? monophobia (ov)`
- `13. superbia (ov)` and `deadmau5 - where's the drop? superbia (ov)`

3. Generally, only 3 imported files have durations that somewhat match any of the tracks in
   the album. How come the rest are so different?

#### Bad matches

| filename                                           | file duration | matched track | matched title | matched duration |
| -------------------------------------------------- | ------------- | ------------- | ------------- | ---------------- |
| **`deadmau5 - where's the drop? superbia (ov)`**   | 2:22          | 5             | avaritia (ov) | 2:58             |
| **`15. strobe (ov)`**                              | 10:36         | 10            | ira (ov)      | 5:29             |
| **`deadmau5 - where's the drop? monophobia (ov)`** | 2:15          | 11            | fn pig (ov)   | 8:32             |
| **`5. avaritia (ov)`**                             | 1:20          | 14            | caritas (ov)  | 1:38             |

This is why I love beets: seeing it I'd firstly question whether I'm trying to import the correct files. 
For example, you may notice that `15. strobe (ov)` duration matches duration of the
original (`Strobe`) track, not the orchestral version which is only half as long.

Otherwise, given that titles do not match and if the files do not contain any other metadata, the matches could have been resolved comparing the durations (which is always present in import files):
1. **avaritia** -> **caritas** with 18s difference
2. **superbia** -> **avaritia** with 36s difference
(and then, well, the leftovers, I guess)
3. **strobe** -> **ira**
4. **monophobia** -> **fn pig**

#### Configuration
You can influence matching by adjusting distance weights for each of the tags that
are being matched, for example that's my configuration:

```yaml
match:
  distance_weights:
    album: 4.0
    album_id: 2.0
    albumdisambig: 4.5
    artist: 4.0
    catalognum: 3.0
    country: 1.0
    genre: 1.0
    label: 3.0
    media: 1.0
    mediums: 1.0
    missing_tracks: 0.0
    source: 1.0
    track_artist: 3.0
    track_index: 1.0
    track_id: 1.0
    track_length: 9.0
    track_title: 4.0
    tracks: 4.0
    unmatched_tracks: 0.5
    year: 2.0
    missing_fields: 5.0
```

This reads

* pay most attention to `track_length`
* pay attention to `album`, `artist`, `track_title`, `tracks`, `missing_fields`
* ignore `missing_tracks`, `genre`, `media`, `mediums`, `source`, `track_id`

Once you've ensured the validity of the files you're importing, try adjusting this configuration and see what happens. This of course depends on the data being available in the files' metadata that you're importing.

**Comment by guerda** [comment]
Thank you so much for the detailed analysis! You are on point, I had duplicates which didn't work so well during the import. That is definitely confusing.
For me the biggest learning from my (partly messy) library I tried to import is that I have to deduplicate thoroughly otherwise it will be a pain.
